Add Node::isInclusiveDescendantOf

rniwa opened the pull request:
#### c8f25024db035b47829a88b2694f83273f7234d6
<pre>
Add Node::isInclusiveDescendantOf
<a href="https://bugs.webkit.org/show_bug.cgi?id=292208">https://bugs.webkit.org/show_bug.cgi?id=292208</a>

Reviewed by Anne van Kesteren.

Add isInclusiveDescendantOf and uses it in a few places. Also rename containsIncludingShadowDOM
to isShadowIncludingInclusiveAncestorOf and isDescendantOrShadowDescendantOf to
isShadowIncludingDescendantOf to match the DOM specification&apos;s concepts.

In addition, generalize the optimization in Node::isDescendantOf so that it applies to a shadow
tree as well as a document tree.

Finally, reimplement Node::containsIncludingShadowDOM in terms of isShadowIncludingDescendantOf
to share more code and optimizations.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateCurrentModalNode):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::findTextRange const):
(WebCore::AccessibilityObject::lineForPosition const):
* Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
(WebCore::disconnectSubframes):
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::updateAnimations):
* Source/WebCore/dom/CustomElementDefaultARIA.cpp:
(WebCore::isElementVisible):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::offsetParentForBindings):
(WebCore::Element::elementForAttributeInternal const):
(WebCore::Element::elementsArrayForAttributeInternal const):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::isDescendantOf const):
(WebCore::Node::isShadowIncludingDescendantOf const):
(WebCore::isClosedShadowHiddenUsingSpecDefinition):
(WebCore::Node::isDescendantOrShadowDescendantOf const): Deleted.
(WebCore::Node::contains const): Deleted.
(WebCore::Node::containsIncludingShadowDOM const): Deleted.
* Source/WebCore/dom/Node.h:
(WebCore::Node::isDescendantOf const):
(WebCore::Node::isInclusiveDescendantOf const):
(WebCore::Node::contains const):
(WebCore::Node::isShadowIncludingDescendantOf const):
(WebCore::Node::isShadowIncludingInclusiveAncestorOf const):
(WebCore::Node::isDescendantOrShadowDescendantOf const): Deleted.
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::parentlessNodeMovedToNewDocumentAffectsRange):
* Source/WebCore/dom/SelectorQuery.cpp:
(WebCore::filterRootById):
* Source/WebCore/dom/TreeScopeOrderedMap.cpp:
(WebCore::TreeScopeOrderedMap::add):
* Source/WebCore/dom/TypedElementDescendantIteratorInlines.h:
(WebCore::InclusiveElementDescendantRange&lt;ElementType&gt;::beginAt const):
(WebCore::InclusiveElementDescendantRange&lt;ElementType&gt;::from const):
* Source/WebCore/editing/Editing.cpp:
(WebCore::firstEditablePositionAfterPositionInRoot):
(WebCore::lastEditablePositionBeforePositionInRoot):
(WebCore::updatePositionForNodeRemoval):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::removingNodeRemovesPosition):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::isDescendantOf):
* Source/WebCore/editing/ios/EditorIOS.mm:
(WebCore::Editor::confirmMarkedText):
* Source/WebCore/editing/markup.cpp:
(WebCore::StyledMarkupAccumulator::traverseNodesForSerialization):
* Source/WebCore/html/FormAssociatedCustomElement.cpp:
(WebCore::FormAssociatedCustomElement::setValidity):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::hidePopoverInternal):
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::handlePopoverTargetAction):
* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::HTMLLabelElement::isEventTargetedAtInteractiveDescendants const):
(WebCore::HTMLLabelElement::defaultEventHandler):
* Source/WebCore/page/DOMSelection.cpp:
(WebCore::DOMSelection::setBaseAndExtent):
* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::hitTestResultForDragStart const):
* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::ElementTargetingController::extractTargets):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::nodeWillBeRemoved):
* Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
(WebCore::RenderTextControlSingleLine::nodeAtPoint):
* Source/WebCore/svg/SVGFEImageElement.cpp:
(WebCore::SVGFEImageElement::imageBufferForEffect const):
* Source/WebCore/svg/SVGUseElement.cpp:
(WebCore::SVGUseElement::findTarget const):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::computeElementLayout):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::simulateClickOverFirstMatchingTextInViewportWithUserInteraction):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::scheduleEditorStateUpdateAfterAnimationIfNeeded):

Canonical link: <a href="https://commits.webkit.org/294243@main">https://commits.webkit.org/294243@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/02204f6928b557588b38821a972663ee6582f8a6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/101268 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20930 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/11233 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/106436 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/51895 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/103309 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/21239 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/29425 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/77123 "90 flakes 124 failures") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/34165 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/104275 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/16380 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/91467 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/57470 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/16196 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/9478 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/51243 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/86076 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/9553 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/108772 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/28396 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/20894 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/86101 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/28758 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/87675 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/85659 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/30384 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/8089 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/22495 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/16467 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/28326 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/33597 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/28138 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/31458 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/29696 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/dom/Node.h:
```diff
@@ -444,14 +444,26 @@ class Node : public EventTarget, public CanMakeCheckedPtr<Node> {
+    ALWAYS_INLINE bool isInclusiveDescendantOf(const Node& other) const { return this == &other || isDescendantOf(other); }
+    ALWAYS_INLINE bool isInclusiveDescendantOf(const Node* other) const { return other && isInclusiveDescendantOf(*other); }
+
+    // https://dom.spec.whatwg.org/#concept-tree-ancestor
```
```suggestion
    // https://dom.spec.whatwg.org/#concept-tree-inclusive-ancestor
```

annevk commented on Source/WebCore/dom/Node.h:
```diff
@@ -444,14 +444,26 @@ class Node : public EventTarget, public CanMakeCheckedPtr<Node> {
+    ALWAYS_INLINE bool contains(const Node& other) const { return other.isInclusiveDescendantOf(*this); }
+    ALWAYS_INLINE bool contains(const Node* other) const { return other && contains(*other); }
+
+    // https://dom.spec.whatwg.org/#concept-shadow-including-inclusive-ancestor
```
```suggestion
    // https://dom.spec.whatwg.org/#concept-shadow-including-descendant
```

annevk reviewed (approved):
Looks good, modulo the specification URLs. (This does make me wonder if we want to cast everything in terms of either descendant or ancestor lookup as they are the inverse so there's no real need to have both.)

rniwa committed:
Add Node::isInclusiveDescendantOf
https://bugs.webkit.org/show_bug.cgi?id=292208

Reviewed by Anne van Kesteren.

Add isInclusiveDescendantOf and uses it in a few places. Also rename containsIncludingShadowDOM
to isShadowIncludingInclusiveAncestorOf and isDescendantOrShadowDescendantOf to
isShadowIncludingDescendantOf to match the DOM specification's concepts.

In addition, generalize the optimization in Node::isDescendantOf so that it applies to a shadow
tree as well as a document tree.

Finally, reimplement Node::containsIncludingShadowDOM in terms of isShadowIncludingDescendantOf
to share more code and optimizations.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateCurrentModalNode):
* Source/WebCore/accessibility/AccessibilityObject.cpp:
(WebCore::AccessibilityObject::findTextRange const):
(WebCore::AccessibilityObject::lineForPosition const):
* Source/WebCore/dom/ContainerNodeAlgorithms.cpp:
(WebCore::disconnectSubframes):
* Source/WebCore/dom/ContentVisibilityDocumentState.cpp:
(WebCore::ContentVisibilityDocumentState::updateAnimations):
* Source/WebCore/dom/CustomElementDefaultARIA.cpp:
(WebCore::isElementVisible):
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::offsetParentForBindings):
(WebCore::Element::elementForAttributeInternal const):
(WebCore::Element::elementsArrayForAttributeInternal const):
* Source/WebCore/dom/Node.cpp:
(WebCore::Node::isDescendantOf const):
(WebCore::Node::isShadowIncludingDescendantOf const):
(WebCore::isClosedShadowHiddenUsingSpecDefinition):
(WebCore::Node::isDescendantOrShadowDescendantOf const): Deleted.
(WebCore::Node::contains const): Deleted.
(WebCore::Node::containsIncludingShadowDOM const): Deleted.
* Source/WebCore/dom/Node.h:
(WebCore::Node::isDescendantOf const):
(WebCore::Node::isInclusiveDescendantOf const):
(WebCore::Node::contains const):
(WebCore::Node::isShadowIncludingDescendantOf const):
(WebCore::Node::isShadowIncludingInclusiveAncestorOf const):
(WebCore::Node::isDescendantOrShadowDescendantOf const): Deleted.
* Source/WebCore/dom/Range.cpp:
(WebCore::Range::parentlessNodeMovedToNewDocumentAffectsRange):
* Source/WebCore/dom/SelectorQuery.cpp:
(WebCore::filterRootById):
* Source/WebCore/dom/TreeScopeOrderedMap.cpp:
(WebCore::TreeScopeOrderedMap::add):
* Source/WebCore/dom/TypedElementDescendantIteratorInlines.h:
(WebCore::InclusiveElementDescendantRange<ElementType>::beginAt const):
(WebCore::InclusiveElementDescendantRange<ElementType>::from const):
* Source/WebCore/editing/Editing.cpp:
(WebCore::firstEditablePositionAfterPositionInRoot):
(WebCore::lastEditablePositionBeforePositionInRoot):
(WebCore::updatePositionForNodeRemoval):
* Source/WebCore/editing/FrameSelection.cpp:
(WebCore::removingNodeRemovesPosition):
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::isDescendantOf):
* Source/WebCore/editing/ios/EditorIOS.mm:
(WebCore::Editor::confirmMarkedText):
* Source/WebCore/editing/markup.cpp:
(WebCore::StyledMarkupAccumulator::traverseNodesForSerialization):
* Source/WebCore/html/FormAssociatedCustomElement.cpp:
(WebCore::FormAssociatedCustomElement::setValidity):
* Source/WebCore/html/HTMLElement.cpp:
(WebCore::HTMLElement::hidePopoverInternal):
* Source/WebCore/html/HTMLFormControlElement.cpp:
(WebCore::HTMLFormControlElement::handlePopoverTargetAction):
* Source/WebCore/html/HTMLLabelElement.cpp:
(WebCore::HTMLLabelElement::isEventTargetedAtInteractiveDescendants const):
(WebCore::HTMLLabelElement::defaultEventHandler):
* Source/WebCore/page/DOMSelection.cpp:
(WebCore::DOMSelection::setBaseAndExtent):
* Source/WebCore/page/DragController.cpp:
(WebCore::DragController::hitTestResultForDragStart const):
* Source/WebCore/page/ElementTargetingController.cpp:
(WebCore::ElementTargetingController::extractTargets):
* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::nodeWillBeRemoved):
* Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
(WebCore::RenderTextControlSingleLine::nodeAtPoint):
* Source/WebCore/svg/SVGFEImageElement.cpp:
(WebCore::SVGFEImageElement::imageBufferForEffect const):
* Source/WebCore/svg/SVGUseElement.cpp:
(WebCore::SVGUseElement::findTarget const):
* Source/WebKit/WebProcess/Automation/WebAutomationSessionProxy.cpp:
(WebKit::WebAutomationSessionProxy::computeElementLayout):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::simulateClickOverFirstMatchingTextInViewportWithUserInteraction):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::scheduleEditorStateUpdateAfterAnimationIfNeeded):

Canonical link: https://commits.webkit.org/294243@main
